src/osd/OSD.cc

```diff
@@ -8727,9 +8727,9 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb )
   assert(NULL != sdata);
   sdata->sdata_op_ordering_lock.Lock();
   if (sdata->pqueue->empty()) {
-    sdata->sdata_op_ordering_lock.Unlock();
     osd->cct->get_heartbeat_map()->reset_timeout(hb, 4, 0);
     sdata->sdata_lock.Lock();
```

Comment:
Basically because releasing sdata_op_ordering lock *without* sdata_lock hold allows messenger worker thread (which tries to insert an op) to race with osd_tp thread (which is about to start waiting for the wakeup signal). See the commit message of the original commit for more details: https://github.com/ceph/ceph/commit/bc683385819146f3f6f096ceec97e1226a3cd237